# Maize Tandem Duplicate Evolution
This document describes the detailed procedure that was taken for the various
maize tandem duplicate evolution analyses and data handling.

## Overview
The general categories of the analysis fall into four categories:

- Tandem duplicate identification
- Tandem duplicate summaries
- Divergence date estimation
- Evolutionary hypothesis tests

They will be detailed in the following sections. Commands and scripts will be
given.

## Tandem Duplicate Identification
Tandem duplicates were identified in two ways:

- Filtering SynMap tandem duplicate output
- Sequence similarity of adjacent genes

SynMap identifies tandem duplicates in a slightly strange way: different genes
within a window that have BLAST hits to the same gene in the other genome. It
does not explicitly check sequence similarity of genes within the same genome,
so there is no guarantee that they are true tandem duplicates. It also
identifies tandem duplicates before it clusters BLAST matches into syntenic
blocks, so there will be nonsyntenic duplications identifed this way.

To refine the SynMap clusters and identify additional tandem duplicates, we
calculated an adjusted pairwise similarity metric for each pair of adjacent
genes in B73v4 and PH207v1. Adjusted pairwise simiarity is pairwise similarity,
down-weighted for the number of gaps opened in alignment of the two sequences.
The procedure for calculating this was as follows:

For each pair of adjacent genes:

1. Translate the CDS of the longest transcripts into amino acids.
2. Align them with ClustalOmega, with 10 iterations of refinement.
3. Back-translate the amino acid alignment into nucleotides.
4. Calculate the proportion of gaps and pairwise similarity (1 - `ThetaPi`)
   with `compute` from the `analysis` package. `compute` can interpret shell
   globbing characters (need to be in single quotes):

    ```bash
    compute -i '*.fasta' > Genomewide_Stats.txt
    ```

5. Calculate adjusted pairwise similarity as `(1 - ThetaPi) * (nsites_ug/nsites)`

This procedure was performed for both B73 and PH207. We ran it on both all
adjacent genes genome-wide, and for pairs of genes within SynMap putative
tandem duplicate clusters.

The alignment is implemented in `Scripts/Analysis/Adjacent_Gene_Similarity.py`,
and the plotting is in `Scripts/Plotting/Tandem_Adjusted_Similarity.R`.

A distribution of adjusted pairwise similarity for adjacent B73 genes, adjacent
PH207 genes, B73 putative tandem genes, and PH207 putative tandem genes is
shown below.

![Distribution of Adjusted Pairwise Similarity](Tandem_Adjusted_Similarity.png)

Based on this plot, putative tandem duplicate genes from SynMap and adjacent
genes with at least 0.3 adjusted similarity were kept as true tandem duplicates.

The tandem duplicate clusters were generated with the
`Scripts/Analysis/Generate_Tandem_Clusters.py` script:

```bash
python Generate_Tandem_Clusters.py \
    B73_Tandem_Alignment_Stats.txt \
    B73_Genomewide_Stats.txt > B73_True_Tandem_Clusters.txt
```

Run the same for PH207.

## Tandem Duplicate Summaries
### Counts of Tandem Duplicate Clusters
The first summary of tandem duplicates is a raw count:

| Genotype | Maize1 Clusters | Maize2 Clusters | Nonsyntenic Clusters |
|----------|-----------------|-----------------|----------------------|
| B73      | 938             | 420             | 1,297*               |
| PH207    | 691             | 316             | 1,005*               |

*: Some clusters overlap syntenic and nonsyntenic genes. This happens when a
tandem duplicate happens at the edge of a syntenic block, and the cluster
includes both syntenic and nonsyntenic genes.

Tandem duplicate clusters were assigned into syntenic or nonsyntenic based on
the synteny master file and the B73-PH207 gene key file, generated by ABB. They
are located at `Data/ABB_Synteny/`. The assignment is implemented in
`Scripts/Analysis/Syntenic_Tandem_Cluster_Align_withSingles.py` and in
`Scripts/Analysis/Nonsyntenic_Tandem_Cluster_Align_withSingles.py`. These
scripts also produce alignments of tandem duplicate gene sequences for
divergence date estimation:

```bash
python Syntenic_Tandem_Cluster_Align_withSingles.py Syn_out/ > Syntenic_Cluster_Assignments.txt
python Nonsyntenic_Tandem_Cluster_Align_withSingles.py Nonsyn_out/ > Nonsyntenic_Cluster_Assignments.txt
```

Alignments are written in the directory provided as the argument to the above
Python script.

### Tandem Duplicates Across the Genome
Next, we wanted to generate a plot of the distribution of tandem duplicates
relative to other genomic features, such as subgenome assignment, gene density,
and TE density. To do this, we need to generate files that give the densities
of these genomic features in windows across the genome:

```bash
python B73_DNA_TE_Gene_Density.py B73_Genes.gff \
    B73_DNA_TE.gff \
    B73_RNA_TE.gff \
    B73_genome.fa.fai > B73_Genomic_Densities.txt
```

We also need to make a GFF that has the positions of the tandemly duplicated
genes:

```bash
cut -f 2 B73_True_Tandem_Clusters.txt | tr ',' '\n' > B_tandem_genes.txt
grep -f B_tandem_genes.txt B73_Genes.gff > B_tandem_genes.gff
```
Then, plot the data with `Scripts/Plotting/B73_Chr2_Plot.R`. Edit the scipt to
potint to the syntenic blocks assignments from ABB, the tandem gene GFF, and the
genomic feature density file. An example for chromosome 2 is shown below:

![Tandem duplications across chromosome 2](B73-PH207_Chr2_Tandem.png)

The top panel shows B73 Chr2, and the bottom panel shows PH207 Chr2. The green
shading is maize1, and the blue shading is maize2. Purple tick marks are
tandem duplicate genes. The black line shows genes per Mb, the red line shows
RNA TEs per Mb, and the orange line shows DNA TEs per MB. Structural annotation
for the TEs in PH207 is not yet available.

It looks like tandem duplications occur where there are genes. That is, it does
not seem like tandem duplications are more common in one subgenome than the
other, nor do they enriched outside of syntenic blocks or around any broad class
of transposable elements.

### Tandem Duplicate Homology
The next summary we want to make relates to the homology of tandem duplicate
clusters between B73 and PH207. To do this, we use homology assignments for
genes between B73 and PH207, and ask if they are duplicated in B73 or PH207.
This is implemented in 
`Scripts/Data_Handling/Identify_Homolgous_Tandems_wSynteny.py`. The first
argument is a file generated by ABB that lists the Ancestral, B1, B2, P1, and P2
genes.

```bash
python Identify_Homologous_Tandems_wSynteny.py \
    b73-ph207-synteny-bt2corrected-nah-coord.txt \
    B73_True_Tandem_Clusters.txt \
    PH207_True_Tandem_Clusters.txt > B_P_Homologous_Tandems.txt
```

Then, plot with `Scripts/Plotting/Tandem_Homology.R`. A heatmap of the
homologous cluster sizes is shown below:

![Homologous tandem cluster sizes](B73_PH207_Homologous_Sizes.png)

There is substantial off-diagonal heat, but it still seems like most of the
time, when a gene is in a tandem duplicate cluster in one genotype, it is in a
cluster of similar size in the other genotype.

## Divergence Date Estimation
Divergence dates were estimated with BEAST. We used the alignments generated by
the `Scripts/Analysis/Syntenic_Tandem_Cluster_Align_withSingles.py` script (and
its nonsyntenic counterpart). Because BEAST is a gene-by-gene program that uses
an XML control file, we need to generate a separate XML file for each tandem
duplicate that we want to analyze. The script to do this is
`Scripts/Data_Handling/Generate_BEAST_XML.py`:

```bash
mkdir Syntenic_BEAST_XML
cd Syn_out
for i in $(*.fasta)
do
    python Generate_BEAST_XML.py ${i} > ../Syntenic_BEAST_XML/${i/.fasta/.xml}
done
```

The same was done for the nonsyntenic duplicates. Note that for these analyses,
the homologous B73 and PH207 tandem duplicate clusters were combined into the
same alignment.

The XML file specifies the parameters of the BEAST analysis. The parameters were
chosen by analyzing several tandem clusters by hand, and adjusting the paramters
to improve the model fit. We chose the following parameters:

- GTR+Gamma nucleotide substitution model
- Estimate the transition probabilities and equilibrium base frequencies
- Random local clock to allow for rate variation among branches
- Prior on maize1-maize2 divergence is ~N(11.9, 1) and monophyletic
- 10,000,000 MCMC steps

BEAST was then run on the MSI servers, using the `Scripts/Job_Scripts/BEAST.job`
script.

Output trees were annotated with time to most recent common ancestry (TMRCA,
"node height") and variances for each estimate with the `TreeAnnotator`
program, pacakged with BEAST. Annotation was done on the MSI servers with
the `Scripts/Job_Scripts/TreeAnnotator.job` script.

The directory of trees was downloaded and parsed locally to obtain distributions
of estimated duplication times. The `Scrits/Analysis/Syntenic_Tandem_Ages.py`
and `Scripts/Analysis/Nonsyntenic_Tandem_Ages.py` will parse the annotated
trees from syntenic duplications and nonsyntenic duplications and print the
ages:

```bash
python Syntenic_Tandem_Ages.py Syn_Annotated_Trees/ > Syntenic_Duplicate_Ages.txt
python Nonsyntenic_Tandem_Ages.py Nonsyn_Annotated_Trees/ > Nonsyntenic_Duplicate_Ages.txt
```

The dates were estimated by finding the time to common ancestry for adjacent
genes within tandem duplicate clusters. While it would be best to know the
number of duplication events, this information is impossible to know. There
must be at most the number of genes - 1 total duplication events, however. In
the ages files generated by the commands above, the ages are reported for
neighboring tandem duplicates. For example:

```
Gen1,Gene2,Gene3    Age1,Age2
```

means that the TMRCA for Gene1 and Gene2 is Age1, and the TMRCA of Gene2 and
Gene3 is Age2.

When a duplication is shared between B73 and PH207, we assume that dates that
match exactly between the two genotypes represent one duplication event that
predates the divergence of the genotypes. This may not necessary be a correct
assumption, because gene duplications likely do not follow the infinite sites
mutation model. I.e., duplications can (and do) happen multiple times at the
same locus, and identity by state is not necessarily identity by descent. We
move foward knowing this limitation.

We also assume that the genes within a tandem duplication are evoloving along
truly separate trajectories. That is, we assume that the genes are truly
divergent, and there is not recombination among them. We know that gene
conversion does happen between genes in tandem arrays, however. This may bias
our results to make duplications look younger than they actually are, because
it would generate more similar sequences than without gene conversion. Knowing
this, too, we continue.

A distribution of the estimated divergence times is showen below:

![Distribution of estimated tandem ages](Tandem_Ages_Clean.png)

Duplications that are shared between B73 Maize1 and PH207 Maize1 are shown in
solid green. Duplications that are shared between B73 Maize2 and PH207 Maize2
are shown in solid blue. Duplications that are private to B73 Maize1 or PH207
Maize1 are shown in dashed green. Duplications that are private to B73 Maize2
or PH207 Maize2 are shown in dashed blue. Nonsyntenic duplications are shown in
solid red.

As expected, duplications that are shared between the subgenomes tend to be
older than duplications that are pricate to a subgenome. Nonsyntenic tandem
duplicates show a high proportion of recent duplicates.

## Evolutionary Hypothesis Tests
We are interested in whether tandem duplicates show different evolutionary rates
than non-tandem genes in maize. To test this, we compare sequence substitution
rates among maize tandem duplications, maize genes generally, and grass
orthologues.

### Identifying Grass Orthologues
We identified orthologues among publicly available grass genomes from
Phytozome V12 and Ensembl Plants V34 with Orthofinder. The amino acid sequences
from the following genome assmeblies were used as input:

| Species                   | Assembly Ver. | Anno. Ver. | Source            |
|---------------------------|---------------|------------|-------------------|
| *Aegilops tauschii*       | ASM34733v1    | ASM34733v1 | Ensembl Plants 34 |
| *Brachypodium distachyon* | 3.1           | 3.1        | Phytozome V12     |
| *Hordeum vulgare*         | ASM32608v1    | ASM32608v1 | Ensembl Plants 34 |
| *Leersia perrieri*        | Lperr_V1.4    | Lperr_V1.4 | Ensembl Plants 34 |
| *Oropetium thomaeum*      | 1.0           | 1.0        | Phytozome V12     |
| *Oryza sativa*            | IRGSP-1.0     | IRGSP-1.0  | Ensembl Plants 34 |
| *Panicum hallii*          | 2.0           | 2.0        | Phytozome V12     |
| *Panicum virgatum*        | 1.1           | 1.1        | Phytozome V12     |
| *Phyllostachys edulis*    | 1.0           | 1.0        | BambooGDB         |
| *Setaria italica*         | 2.2           | 2.2        | Phytozome V12     |
| *Sorghum bicolor*         | 3.1           | 3.1        | Phytozome V12     |
| *Triticum aestivum*       | TGACv1        | TGACv1     | Ensembl Plants 34 |
| *Triticum urartu*         | ASM34745v1    | ASM34745v1 | Ensembl Plants 34 |

Orthofinder was run with the default 'dendroblast' orthologue search method,
and default parameters for MCL clustering. Amino acid sequences from B73 and
PH207 were kept in separate files. Be sure to download both the amino acid
sequences and the nucleotide sequences. The amino acids sequences will be used
to identify orthologues, and the nucleotide sequences will be used in alignment
and tree building. The amino acid files are saved in `Ortho_Base` and the
nucleotide files are saved in `Ortho_Nuc`.

Orthologue searching was run on MSI, and the job file is given in
`Scripts/Jobs/Orthofinder.job`.

### Generating Trees
The next step in generating the input files is to generate nucleotide sequence
alignments and phylogenetic trees that relate the sequences in each orthologous
group. Because Orthofidner uses a modification of BLAST scores, it does not
produce any alignments nor trees. We must do this on our own.

First, retrieve the sequences that are part of each orthogroup. We use the CSV
orthogroup output file, and the directory of amino acid input files for
Orthofinder. The `Scripts/Data_Handling/Generate_Orthogroup_Seqs.py` script
will do this:

```bash
mkdir Orthogroup_Seqs/
python Generate_Orthogroup_Seqs.py \
    Orthogroups.csv.gz \
    /path/to/Ortho_Base/ \
    Orthogroup_Seqs/
```

Note that this script requires `samtools` to be available, because it uses
`samtools faidx` to retrieve sequences from the amino acid FASTA files.

Then, we align each orthogroup with ClustalOmega:

```bash
mkdir Aligned/
cd Orthogroup_Seqs/
for i in *.fa
do
    clustalo -v -i ${i} -t Protein --full-iter --iter=10 --out ../Aligned/${i/.fa/_Aligned.fa}
done
```

This can take a long time, so we do it on MSI.

Then, the orthogroup alignments need to be backtranslated from amino acids to
nucleotides. The script `Scripts/Data_Handling/Backtranslate_Orthogroup.py`
will do this:

```bash
mkdir Backtranslated/
cd Aligned/
for i in *Aligned.fa
do
    python Backtranslate_Orthogroup.py /path/to/Ortho_Nuc ${i} > ../Backtranslated/${i/_Aligned/_Backtranslated}
done
```

This step is relatively fast.

Finally, we use RAxML to estimate a tree of the sequences in the aligned and
backtranslated orthogroup. We use the following parameters:

- Default rapid hill-climbing search algorithm (`-f d`)
- GTR+Gamma nucleotide substitution model (`-m GTRGAMMAX`)

```bash
mkdir Trees/
cd Backtranslated/
for i in *Backtranslated.fa
do
    raxml -f d -m GTRGAMMAX -n ${i/_Backtranslated.fa/} -s ${i} -p 123 -w /full/path/to/Trees
done
```

### CodeML (PAML) Tests
## Software Versions and Environment
