# Maize Tandem Duplicate Evolution
This document describes the detailed procedure that was taken for the various
maize tandem duplicate evolution analyses and data handling.

## Overview
The general categories of the analysis fall into four categories:

- Tandem duplicate identification
- Tandem duplicate summaries
- Divergence date estimation
- Evolutionary hypothesis tests

They will be detailed in the following sections. Commands and scripts will be
given.

## Tandem Duplicate Identification
Tandem duplicates were identified in two ways:

- Filtering SynMap tandem duplicate output
- Sequence similarity of adjacent genes

SynMap identifies tandem duplicates in a slightly strange way: different genes
within a window that have BLAST hits to the same gene in the other genome. It
does not explicitly check sequence similarity of genes within the same genome,
so there is no guarantee that they are true tandem duplicates. It also
identifies tandem duplicates before it clusters BLAST matches into syntenic
blocks, so there will be nonsyntenic duplications identifed this way.

To refine the SynMap clusters and identify additional tandem duplicates, we
calculated an adjusted pairwise similarity metric for each pair of adjacent
genes in B73v4 and PH207v1. Adjusted pairwise simiarity is pairwise similarity,
down-weighted for the number of gaps opened in alignment of the two sequences.
The procedure for calculating this was as follows:

For each pair of adjacent genes:

1. Translate the CDS of the longest transcripts into amino acids.
2. Align them with ClustalOmega, with 10 iterations of refinement.
3. Back-translate the amino acid alignment into nucleotides.
4. Calculate the proportion of gaps and pairwise similarity (1 - `ThetaPi`)
   with `compute` from the `analysis` package. `compute` can interpret shell
   globbing characters (need to be in single quotes):

    ```bash
    compute -i '*.fasta' > Genomewide_Stats.txt
    ```

5. Calculate adjusted pairwise similarity as `(1 - ThetaPi) * (nsites_ug/nsites)`

This procedure was performed for both B73 and PH207. We ran it on both all
adjacent genes genome-wide, and for pairs of genes within SynMap putative
tandem duplicate clusters.

The alignment is implemented in `Scripts/Analysis/Adjacent_Gene_Similarity.py`,
and the plotting is in `Scripts/Plotting/Tandem_Adjusted_Similarity.R`.

A distribution of adjusted pairwise similarity for adjacent B73 genes, adjacent
PH207 genes, B73 putative tandem genes, and PH207 putative tandem genes is
shown below.

![Distribution of Adjusted Pairwise Similarity](Tandem_Adjusted_Similarity.png)

Based on this plot, putative tandem duplicate genes from SynMap and adjacent
genes with at least 0.3 adjusted similarity were kept as true tandem duplicates.

## Tandem Duplicate Summaries
### Counts of Tandem Duplicate Clusters
The first summary of tandem duplicates is a raw count:

| Genotype | Maize1 Clusters | Maize2 Clusters | Nonsyntenic Clusters |
|----------|-----------------|-----------------|----------------------|
| B73      | 938             | 420             | 1,297*               |
| PH207    | 691             | 316             | 1,005*               |

*: Some clusters overlap syntenic and nonsyntenic genes. This happens when a
tandem duplicate happens at the edge of a syntenic block, and the cluster
includes both syntenic and nonsyntenic genes.

Tandem duplicate clusters were assigned into syntenic or nonsyntenic based on
the synteny master file and the B73-PH207 gene key file, generated by ABB. They
are located at `Data/ABB_Synteny/`. The assignment is implemented in
`Scripts/Analysis/Syntenic_Tandem_Cluster_Align_withSingles` and in
`Scripts/Analysis/Nonsyntenic_Tandem_Cluster_Align_withSingles`. These scripts
also produce alignments of tandem duplicate gene sequences for divergence
date estimation:

```bash
python Syntenic_Tandem_Cluster_Align_withSingles.py Syn_out/ > Syntenic_Cluster_Assignments.txt
python Nonsyntenic_Tandem_Cluster_Align_withSingles.py Nonsyn_out/ > Nonsyntenic_Cluster_Assignments.txt
```

Alignments are written in the directory provided as the argument to the above
Python script.

### Tandem Duplicates Across the Genome
Next, we wanted to generate a plot of the distribution of tandem duplicates
relative to other genomic features, such as subgenome assignment, gene density,
and TE density. To do this, we need to generate files that give the densities
of these genomic features in windows across the genome:

```bash
python B73_DNA_TE_Gene_Density.py B73_Genes.gff \
    B73_DNA_TE.gff \
    B73_RNA_TE.gff \
    B73_genome.fa.fai > B73_Genomic_Densities.txt
```

We also need to make a GFF that has the positions of the tandemly duplicated
genes:

```bash
cut -f 2 B73_True_Tandem_Clusters.txt | tr ',' '\n' > B_tandem_genes.txt
grep -f B_tandem_genes.txt B73_Genes.gff > B_tandem_genes.gff
```
Then, plot the data with `Scripts/Plotting/B73_Chr2_Plot.R`. Edit the scipt to
potint to the syntenic blocks assignments from ABB, the tandem gene GFF, and the
genomic feature density file. An example for chromosome 2 is shown below:

![Tandem duplications across chromosome 2](B73-PH207_Chr2_Tandem.png)

The top panel shows B73 Chr2, and the bottom panel shows PH207 Chr2. The green
shading is maize1, and the blue shading is maize2. Purple tick marks are
tandem duplicate genes. The black line shows genes per Mb, the red line shows
RNA TEs per Mb, and the orange line shows DNA TEs per MB. Structural annotation
for the TEs in PH207 is not yet available.

### Tandem Duplicate Homology
The next summary we want to make relates to the homology of tandem duplicate
clusters between B73 and PH207. To do this, we use homology assignments for
genes between B73 and PH207, and ask if they are duplicated in B73 or PH207.
This is implemented in 
`Scripts/Data_Handling/Identify_Homolgous_Tandems_wSynteny.py`. The first
argument is a file generated by ABB that lists the Ancestral, B1, B2, P1, and P2
genes.

```bash
python Identify_Homologous_Tandems_wSynteny.py \
    b73-ph207-synteny-bt2corrected-nah-coord.txt \
    B73_True_Tandem_Clusters.txt \
    PH207_True_Tandem_Clusters.txt > B_P_Homologous_Tandems.txt
```

Then, plot with `Scripts/Plotting/Tandem_Homology.R`. A heatmap of the
homologous cluster sizes is shown below:

![Homologous tandem cluster sizes](B73_PH207_Homologous_Sizes.png)

## Divergence Date Estimation
 Divergence dates were estimated with 
## Evolutionary Hypothesis Tests
