<html>

<head>
<title>Zm_Tandem_Duplicates</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<style type="text/css">
	body {background-color: #bac0c7}
    p.binderItem {margin: 10.0px 0.0px 0.0px 05.0px; font-family:Cochin, Times, Courier, Arial, serif; font-size:14.0px;}
    .page {border: 1px solid #727272; background: #fff}
    hr {
      border-top: 1px dashed #000;
      border-bottom: 0px solid #fff;
      color: #fff;
      background-color: #fff;
      height: 0px;
  </style>
</head>

<body>

<table border="0" width="100%" cellspacing="3">
<tr>
<td>

<table class="page" width="100%" cellspacing="10" cellpadding="2">
<tr>
<td valign="top">

<ul>
<li>
<p class="binderItem"><strong>Abstract</strong><br/>Gene duplication is a mutation type that has high potential to affect phenotypes. In addition to altering gene dosage, duplications may provide a substrate for further phenotypic change by allowing for rapid protein function divergence. However, not all gene duplications result in novel functional proteins; many duplicated genes lose function through pseudogenization. Here, we investigate the frequency of tandem gene duplication and the relative rates of the tandem duplicate fates in two inbred ...</p>
</li>
<li>
<p class="binderItem"><strong>Introduction</strong></p>
</li>
<hr/>
<ul>
<li>
<p class="binderItem"><strong>How Duplications Arise</strong><br/>Gene duplications provide a mechanism through which functional novelty may arise. Many protein coding genes in eukaryotes are part of large families of genes with related function, consistent with origins in gene duplication (Rubin et al. 2000). Duplicated genes may either be tandem (proximal to each other), or dispersed (separated in the genome). Dispersed duplicate genes may functionally diverge because the two copies may be subject to different regulatory pressures. Conversely, genes that are...</p>
</li>
<li>
<p class="binderItem"><strong>Evolutionary Fates of Duplicates</strong><br/>When a gene duplicates, it potentially generates a phenotypic impact through alteration of gene product dosage. In many cases, the sudden change of gene product concentration has deleterious  effects on the physiology of the organism, and will be selected against. In some cases, however, the increased gene expression may be beneficial, and there will be selection to maintain the duplication, e.g., tandem duplications conferring soybean cyst nematode resistance at Rhg1 (Cook et al. 2012).<br/>
In the ...</p>
</li>
<li>
<p class="binderItem"><strong>Technological Improvements</strong><br/>Recent improvements in DNA sequencing technology allow for the resolution of duplicated sequences. Reads are now long enough to span repetitive elements, where shorter reads would either miss or collapse repetitive sequences. In maize, the latest B73 reference genome was assembled with long read sequencing, which allows for improved resolution of repetitive sequences than previous assemblies. Additionally, a de noco genome assembly of PH207, another maize inbred line, has become recently availab...</p>
</li>
<li>
<p class="binderItem"><strong>Major Questions</strong><br/>The major questions we aim to address in this study are 1) How many genes are tandemly duplicated in the B73 and PH207 reference genomes, and where do they occur? 2) How many tandem duplicate genes show evidence of functional divergence, at the DNA sequence or expression levels? 3) What is the age distribution of tandem duplications, and is there a relationship between their age and functional outcome? To answer these questions, we compared the B73 and PH207 reference genomes to the outgroup spe...</p>
</li>
</ul>
<hr/>
<li>
<p class="binderItem"><strong>Materials and Methods</strong></p>
</li>
<hr/>
<ul>
<li>
<p class="binderItem"><strong>Genome Sequences and Databases</strong><br/>The B73v4 and PH207v1 reference sequences, including gene annotations, were downloaded from Phytozome V12 (link). Structural annotations of transposable elements in the B73 and PH207 reference sequences were provided by Michelle Stitzer. Ancestral states were estimated with the Sorghum bicolor and Oryza sativa reference sequences, available from Ensembl Plants (link). </p>
</li>
<li>
<p class="binderItem"><strong>Tandem Duplicate Identification</strong><br/>Tandem duplicate genes were identified with the ‘blast2raw’ script from the CoGe pipeline (cite, link). A window size of 15 genes was used to identify potential tandem duplicates, with a C-value of 0.1. Tandem duplicate genes in regions that are syntenic with Sorghum bicolor were identified by comparisons of each maize reference sequence to the S. bicolor v3 reference sequence. Tandem duplicate genes in non-syntenic regions were identified by comparison of the B73 reference sequence to the PH207...</p>
</li>
<li>
<p class="binderItem"><strong>Orthologue Identification And Evolutionary Hypothesis Testing</strong><br/>Orthologous gene groups were identified with Orthofinder v. 1.1.5 (cite, link). Briefly, Orthofinder identifies orthologous groups of genes by reciprocal protein BLAST searches followed by MCL clustering. Orthofinder was run with the default E-value threshold of 0.001, and the DendroBLAST algorithm to infer orthologous groups. Orthofinder’s dependencies were satisfied with MCL v. 10.201 (cite), BLAST+ v. 2.2.29 (cite), MAFFT v. 7.305 (cite), FastTree v. 2.1.8 (cite), and DLCpar v. 1.0 (cite). Th...</p>
</li>
<li>
<p class="binderItem"><strong>Duplication Dating</strong><br/>Ages of tandem duplicate genes were estimated using the synonymous nucleotide substitution rate. Full coding sequence from each gene of each putative tandem duplicate cluster and an inferred ancestral gene sequence from Sorghum bicolor or Oryza sativa were translated to amino acids, then iteratively aligned with Clustal Omega (cite). Clustal Omega was run with a maximum of 10 iterations and with the full distance matrix used for each clustering iteration. Alignments were then backtranslated to n...</p>
</li>
</ul>
<hr/>
<li>
<p class="binderItem"><strong>Results</strong></p>
</li>
<hr/>
<ul>
<li>
<p class="binderItem"><strong>Tandem Duplicate Gene Clusters</strong><br/>We identified a total of 700 tandem duplicate gene clusters in B73, and 600 tandem duplicate gene clusters in PH207. These clusters contained ___ (__% of protein coding genes in B73) and ___ (__)% of protein coding genes in PH207) genes, respectively. Of the tandem duplicate clusters identified, 76 appear to be part of cassettes in B73, and 179 appear to be part of cassettes in PH207. Tandem duplicate genes were distributed randomly across the genome (Figure 1).<br/>
With our thresholds of 75% identi...</p>
</li>
<li>
<p class="binderItem"><strong>Functional Roles of Tandem Duplicates</strong></p>
</li>
</ul>
<hr/>
<li>
<p class="binderItem"><strong>Discussion</strong></p>
</li>
</ul>

</td>
<td width="8">
</td>
</tr>
</table>

</td>
</tr>
</table>

</body>
</html>